### ISSUE TITLE
Sample decoding switches to Multiplex decoding

### ISSUE THREAD

**Issue by cliffbueno** [biosails/pheniqs#41]
Hello,
In my config file I tried to set up Sample decoding according to all of the examples I have seen. When I validate the file, after the Output transform section, there should be a "Sample decoding" section, but for me there is a "Mutliplex decoding" (yes, it says Mutliplex not Multiplex, I wonder if anyone has ever noticed that :) ).

My first step in troubleshooting was to try an example file instead of my own file. I found in another issue (#32)  you posted a corrected config file as well as the correct validation output. The validation output you posted correctly shows a "Sample decoding" section, with info for each barcode and output. However, when I validate the same file, I see a "Mutliplex decoding" section instead, with very limited output and no barcode info and file output info.

I used the same code, where original.json is copied from your response in #32
pheniqs mux --config original.json -F fastq --compression gz --validate

I'm using a conda install of pheniqs 2.0.4

Any ideas? Thanks for your help.
Cliff

**Comment by cliffbueno** [comment]
Hi Lior,
Sure - here is the "original.json" from Issue 32 as well as "original_compiled.json" which is my output from `pheniqs mux --config original.json --compile`
As well as my own "pheniqs_config_mdd.json" (159 barcodes, and I wanted to try mdd with allowing 1 mismatch). For my own project, it's 12 bp Illumina barcodes but Nanopore sequencing. 

For me, when I run the validate step, the output shows a "Mutliplex decoding" section with no barcode info. I also noticed some other difference compared to your output in #32 in terms of the environment, assembly instruction, and not showing compression in the Input feed.

For original.json, it validation output looks like this.

`conda activate pheniqs_env`
`pheniqs --version`
pheniqs version 2.0.4
`pheniqs mux --config original.json -F fastq --compression gz --validate`

Environment 

    Base input URL                              .
    Base output URL                             .
    Platform                                    ILLUMINA
    Quality tracking                            disabled
    Include non PF reads                        disabled
    Input Phred offset                          33
    Output Phred offset                         33
    Leading segment index                       0
    Default output format                       fastq
    Default output compression                  gz
    Feed buffer capacity                        2048
    Threads                                     32

Input 

    Input segment cardinality                   4

    Input segment No.0 : ./JMSNRSCF_S1_L001_R1_001.fastq.gz
    Input segment No.1 : ./JMSNRSCF_S1_L001_I1_001.fastq.gz
    Input segment No.2 : ./JMSNRSCF_S1_L001_I2_001.fastq.gz
    Input segment No.3 : ./JMSNRSCF_S1_L001_R2_001.fastq.gz

    Input feed No.0
        Type : fastq
        Resolution : 1
        Phred offset : 33
        Platform : ILLUMINA
        Buffer capacity : 2048
        URL : ./JMSNRSCF_S1_L001_R1_001.fastq.gz

    Input feed No.1
        Type : fastq
        Resolution : 1
        Phred offset : 33
        Platform : ILLUMINA
        Buffer capacity : 2048
        URL : ./JMSNRSCF_S1_L001_I1_001.fastq.gz

    Input feed No.2
        Type : fastq
        Resolution : 1
        Phred offset : 33
        Platform : ILLUMINA
        Buffer capacity : 2048
        URL : ./JMSNRSCF_S1_L001_I2_001.fastq.gz

    Input feed No.3
        Type : fastq
        Resolution : 1
        Phred offset : 33
        Platform : ILLUMINA
        Buffer capacity : 2048
        URL : ./JMSNRSCF_S1_L001_R2_001.fastq.gz

Output transform

    Output segment cardinality                  4

    Token No.0
        Length        variable
        Pattern       0::
        Description   cycles 0 to end of input segment 0

    Token No.1
        Length        variable
        Pattern       1::
        Description   cycles 0 to end of input segment 1

    Token No.2
        Length        variable
        Pattern       2::
        Description   cycles 0 to end of input segment 2

    Token No.3
        Length        variable
        Pattern       3::
        Description   cycles 0 to end of input segment 3

    Assembly instruction
        Append token 0 of input segment 0 to output segment 0
        Append token 1 of input segment 1 to output segment 1
        Append token 2 of input segment 2 to output segment 2
        Append token 3 of input segment 3 to output segment 3

Mutliplex decoding

    Decoding algorithm                          transparent

    Barcode undetermined
        ID : JMSNRSCF:undetermined
        PU : JMSNRSCF:undetermined
        PL : ILLUMINA
        PM : NovaSeq_2x250_UCDavis_run064
        Segment No.0  : /dev/stdout
        Segment No.1  : /dev/stdout
        Segment No.2  : /dev/stdout
        Segment No.3  : /dev/stdout

    Output feed No.0
        Type : fastq
        Resolution : 4
        Phred offset : 33
        Platform : ILLUMINA
        Buffer capacity : 8192
        URL : /dev/stdout

Thanks for your help,
Cliff
[original_compiled.json](https://github.com/user-attachments/files/17478976/original_compiled.json)
[original.json](https://github.com/user-attachments/files/17478977/original.json)
[pheniqs_config_mdd.json](https://github.com/user-attachments/files/17478978/pheniqs_config_mdd.json)

**Comment by moonwatcher** [comment]
Sorry it took me some time to realize you are using 2.0.4. there were changes upstream to how this is all handled. I think 2.0.4 only allowed to set individual outputs for the sample barcodes and later this was replaced with a "muliplexing decoder" directive that allows to define which of the decoders will control output multiplexing. configuration files will not be portable between those versions and pheniqs generally ignores directives in the config file that it does not expect, to allow you to add your own values as comments and documentation. Can you try this with the latest version that is in conda?

**Comment by cliffbueno** [comment]
Oh, thanks for pointing that out, because I thought 2.0.4 was the latest version, because I just installed this last week following the Installation guidelines.
```
conda config --add channels defaults
conda config --add channels conda-forge
conda config --add channels bioconda
conda create -n pheniqs_env pheniqs
```
I just went to the anaconda website though and it looks like there is a version 2.1.0 so I will see if `conda install bioconda::pheniqs` will work.

**Comment by cliffbueno** [comment]
Ok it was not so simple. Updated conda and tried again but it was still installing 2.0.4. I had to troubleshoot python versions and other package versions.

This ended up working to get me pheniqs 2.1.0: `mamba create -n pheniqs python=3.9 conda-forge::libdeflate=1.21.0 bioconda::htslib=1.21.0 bioconda::pheniqs=2.1.0`

And then after that, `pheniqs mux --config original.json -F fastq --compression gz --validate`, yielded the correct output.

Thanks for your help!

**Comment by moonwatcher** [comment]
Glad that is sorted, there have been several iterations on how to make the configuration as flexible as possible.

Btw, you really should give pamld a spin, it should be able to handle errors much better. At least with substitution it has less false negatives and less false positives, so it's a no brainer. 

ONT has more indels so I am not sure how it would do there but it won't be worse than mdd.

**Comment by cliffbueno** [comment]
Okay. I'm not really sure which parameters to put for that in terms of noise and confidence_threshold though. I read the paragraph about that in the documentation, but I'm not sure what might be best for my situation.

We have 390 bp inserts, with 12 bp barcodes. ONT minION run with average quality score of 12. I'm looking to classify about 54% of reads with my barcode set (the other 46% should be from a different set of samples that I am not interested in). Given this info, do you have any suggestions for approximate values to try for noise and confidence threshold?

**Comment by moonwatcher** [comment]
ok. The main point is that assuming you have a decently multithreaded processor pheniqs will be blazing fast so you can really experiment. It will pretty much saturate any amount of cores you give it assuming IO can keep up. 

Like everything in statistics 0.95 is a good initial value for confidence_threshold, it's basically a threshold for an acceptable posterior probability. going to 0.99 will mean a lot more false negatives (more reads being tossed out) and maybe a few less false positives. Because ONT has qualities that are so much lower than Illumina, it's probably worth trying lower values to see what happens, like 0.9 and 0.8. PAMLD will not know how to handle an indel so i would say with almost absolute certainty that a read with an indel would end up being classified as noise (it would take a serious fluke for a shift to make one barcode look like another). MDD just ignores the qualities so for Illumina, where the error rate is 0.1 to 0.01% you can argue that PAMLD is an "improvement" over MDD. Phred 12 is 0.0630957344 error probability, or 6.3 out of 100 bases, I would say that  for ONT error rates ignoring the quality means you are just guessing ;)

Noise is important and should be around the % of reads that you expect to not get classified. or the chance that a barcode is not meant to match one of the values you provided. if you already know you want to ignore 0.46 of the data that would be the lowest value I would set noise to. There might be more noise (Illumina always has some spiked in PhiX but I don't think that is used in ONT) so you might want a number a little higher, but this is just a prior. as long as its in the ball park it would be fine. Just don't set it to 0...

are the rest of the barcodes evenly distributed? you can potentially try to do a second run. Pheniqs can use the statistics from a run to recalibrate the priors on a connective run. All you need to do is specify `--prior` and set it to a path to a json file, for instance `--prior calibrated_config.json` and at the end of the first run pheniqs will write a config similar to the one you started with but with adjusted priors inferred from statistics collected in the first run. 

I don't have much experience with ONT but if you are willing to experiment (and have access to the computational capacity) I can help you navigate this :)

**Comment by moonwatcher** [comment]
BTW, looking over your config, MDD does not have an `allow_mismatch` property. the correct directive is `distance tolerance` and it's an array of integers, same size as the number of segments in the barcode. You specify the tolerance for each segment independently. if your barcode has only one segment that would be: 
```json
"distance tolerance": [ 1 ]
```

https://github.com/biosails/pheniqs/blob/ea31a77e803f5326d92fc5ffba92d1b65e63e657/configuration.json#L822

**Comment by cliffbueno** [comment]
Thank you very much for these insights, very helpful. I'm working on a server and it runs in a few minutes so I will play around with this some.

**Comment by moonwatcher** [comment]
When we wrote the paper pheniqs got through a lane of NovaSeq with over 2B reads in about an hour on one of the cluster nodes (some Xeon with 24 cores I think, details are in the paper). the surprising thing is that MDD and PAMLD run at about the same time even though PAMLD does a huge amount of arithmetic and MDD is just a string comparison. its basically mostly IO bound...

**Comment by cliffbueno** [comment]
This is running on 15 million sequences in 2 minutes on 32 cores!
So far MDD is getting too few sequences and PAMLD is getting too many.
I'm going to get the full mapping file and just process them all together so I don't have to worry about the 46% issue.
I'm also going to get some barcodes that weren't sequenced, to serve as a demultiplexing control.

**Comment by cliffbueno** [comment]
Okay, getting back to this. I now have all samples/barcodes for the run. So the only reads that should not be assigned barcodes are the Device Calibration Strand (DCS) control reads, which should be about 1-2% of the reads. I've tried a couple different parameters so far, here are the results:

mdd 1 mismatch: 22.80% classified
mdd 2 mismatch: 24.59% classified
pamld 0.05 noise, 0.95 conf. thresh: 96.15% classified
pamld 0.05 noise, 0.99 conf. thresh: 96.15% classified

Changing confidence from 0.95 to 0.99 had no effect. And I know there are way to many errors with these pamld settings as tons of reads are getting assigned to extraction blanks and no template controls, and I already know from the Illumina MiSeq run on the same library pool that those should have very few reads. The pamld numbers were also almost as high even when I tried before without all of the samples.

Please let me know if you have any ideas for parameters to try!

**Comment by moonwatcher** [comment]
Can I see the log for that pamld run? I think I know what is happening. I assume you have fastq output? If a read passes the noise filter but fails the confidence filter it will still get classified to the class but marked as not passing quality filter. You can see that in SAM. Output or even just filter NPF reads on output. I probably didn't have reference Nanopre reads at the time and didn't implement the PF flag for fastq.

Basically changing the quality threshold won't change how many will be classified just how many PASS Filter.

**Comment by cliffbueno** [comment]
Ok, thanks, here is the log file. Yes, I want fastq output, more specifically fastq.gz. 
[demux_sample_report.json](https://github.com/user-attachments/files/17551217/demux_sample_report.json)

So, how can I implement the PASS Filter?

**Comment by moonwatcher** [comment]
Hey, sorry for the delayed response, this is not my day job :)
So you want to look at the the values in the report that have `pf` in them. That stands for `PASS FILTER`. It's a concept that was initially found in Illumina FASTQ files but has been generalized by the SAM format. Basically it's a way to flag a read as not passing quality filtering. SAM is kind of a superset of FASTQ and has much richer semantics. Pheniqs internally thinks in terms of SAM format records but strips out the additional metadata when it outputs FASTQ. 

Since there is only one flag, there is no way of telling what step in a pipeline actually flagged the read as **not passing quality filtering**. Pheniqs can filter those reads either on input, output or both. its basically two different flags that you can either specify in the config or on the cli: 

```
--no-input-npf   -- Filter incoming QC failed reads
--no-output-npf  -- Filter outgoing QC failed reads
```

in the config they will be in the root element and both default to `false`: 
```json
"filter incoming qc fail": true,
"filter outgoing qc fail": true
```

If we look at one class from your report you can see that the `average classified confidence` is only 26% but the `average pf classified confidence` is 99.84%. there is a page about this on the manual but in a nutshell, the noise filter compares the conditional probability of the best match (just the probability that the best matching barcode would be a match, conditional on the assumption that it is...) to the chance of observing, well, anything... the chance of seeing any of all the possible sequences. if that conditional is lower than the probability of a random sequence that best or not, its not really any better than random and so I more likely to be noise than anything else.  if the match passes that filter it means that its more likely to be a match to the best match than anything else and so its classified to that class. but then comes in the confidence threshold you specify: the posterior probability is the probability of that match being a correct match out of all the options you specified and it needs to be higher than the threshold. if its not, its still classified but marked as **not passing quality filtering** (or NPF for short).  changing the threshold won't really change how many will be classified but **will** change how many pass filtering.  If you don't care to deal with those reads just set `--no-output-npf` on the command line (or in the config).

You can see that for **GCTGCGTATACC** `count` is 96718 but `pf count` is only 24623. Only 25.45% of the reads that were classified to this class passed quality filtering. The noise was estimated at about 13.3%.

```json
    "sample": {
        "average classified confidence": 0.260548444829702,
        "average classified distance": 4.867135487208378,
        "average pf classified confidence": 0.998446856860894,
        "average pf classified distance": 0.396753946858833,
        "classified": [
            {
                "BC": "GCTGCGTATACC",
                "CN": "CU",
                "DT": "2024-10-15",
                "ID": "GCTGCGTATACC",
                "PI": "390",
                "PL": "ONT",
                "PM": "minION",
                "PU": "GCTGCGTATACC",
                "average confidence": 0.254258866321575,
                "average distance": 4.749860418949937,
                "average pf confidence": 0.998717013884991,
                "average pf distance": 0.328879502903789,
                "barcode": [
                    "GCTGCGTATACC"
                ],
                "concentration": 0.005277777777777,
                "count": 96718,
                "estimated concentration": 0.005413415147581,
                "index": 1,
                "low conditional confidence count": 19645,
                "low confidence count": 72095,
                "pf count": 24623,
                "pf fraction": 0.254585495977997,
                "pf pooled classified fraction": 0.006245024181936,
                "pf pooled fraction": 0.006245024181936,
                "pooled classified fraction": 0.006401238339604,
                "pooled fraction": 0.00615451773342
            }
        ],
        "classified count": 15109264,
        "classified fraction": 0.961457362920427,
        "classified pf fraction": 0.260953743345804,
        "count": 15714960,
        "estimated noise": 0.13316346104168,
        "index": 0,
        "low conditional confidence count": 605696,
        "low confidence count": 11166445,
        "pf classified count": 3942819,
        "pf classified fraction": 1.0,
        "pf count": 3942819,
        "pf fraction": 0.250895897921471,
        "unclassified": {
            "CN": "CU",
            "DT": "2024-10-15",
            "ID": "undetermined",
            "PI": "390",
            "PL": "ONT",
            "PM": "minION",
            "PU": "undetermined",
            "count": 605696,
            "index": 0,
            "pf count": 0,
            "pf fraction": 0.0,
            "pf pooled fraction": 0.0,
            "pooled classified fraction": 0.040087723664104,
            "pooled fraction": 0.038542637079572
        }
    }
```

Do you know if the comment field on an ONT fastq record has the PF flag somewhere? if you can show me some consistent description of how metadata is encoded in to the fastq comment field in an ONT fastq I can implement it, so you can see the PF flag on fastq output, but for now you either output SAM/BAM/CRAM or just filter them on output?

**Comment by cliffbueno** [comment]
Thanks for this info! So all I have to do is add --no-input-npf and --no-output-npf to my pheniqs mux command? It seems to have worked. I tried it with 0.05 noise/0.95 confidence threshold, and got 72% (of the reads that had adapters) demultiplexed. This is much less than when I don't include those flags.

I don't think ONT fastq files normally have PF flags. In my fastq, the only thing on the first line is the read ID.

**Comment by moonwatcher** [comment]
Well you don't strictly need both. Once you include --no-output-npf it doesn't really matter if you include --no-input-npf, since pheniqs will never set the PF flag back on once it was turned off. In other words pheniqs can only fail more reads it can't make a failed read "not fail". 

Filtering on input means pheniqs will drop them as they come in so the decoder doesn't even see them, and they won't show up in statistics. Filtering on output means they are just not written to the output but otherwise processed like all the others. It also means it will drop both the ones that pheniqs marked as NPF and no those that were like that on input. If you have a lot of NPF reads on your input than filtering on input will make things a bit faster since they won't even be evaluated.
